### ISSUE TITLE
Eliminate uses of deprecated `FindCUDA.cmake`

### ISSUE THREAD

**Issue by peterbell10** [pytorch/pytorch#76082]
`find_package(CUDA)` is the legacy way of using CMake to build CUDA code and was deprecated in CMake 3.10. #62445 transitioned over to `enable_language(CUDA)` which gives first class support for `CUDA` as a language in CMake.  However, `find_package(CUDA)` is still being used to locate the CUDA toolkit libraries. This is potentially dangerous because we have two different mechanisms searching for the CUDA installation, each with their own configuration variables. So, it would be best to stop using `FindCUDA.cmake` entirely.

The modern CMake equivalent for library targets is `find_package(CUDAToolkit)` which requires CMake 3.17, so we can't assume it's present yet. However, it might also be possible to vendor `FindCUDAToolkit.cmake` and use it anyway.

There are also dependencies in `third_party` which still use `find_package(CUDA)`
- [ ] [gloo](https://github.com/facebookincubator/gloo/blob/c22a5cfba94edf8ea4f53a174d38aa0c629d070f/cmake/Cuda.cmake#L133)
- [ ] [tensorpipe](https://github.com/pytorch/tensorpipe/blob/52791a2fd214b2a9dc5759d36725909c1daa7f2e/tensorpipe/CMakeLists.txt#L237)

**Comment by peterbell10** [comment]
This change would be a lot simpler if we could start using CMake 3.17. From this comment https://github.com/pytorch/pytorch/pull/63660#issuecomment-904666017 it sounds like the criteria is: "does the latest ubuntu LTS release have a cmake that supports it". As it just so happens, the next LTS release is scheduled for tomorrow - April 21st - and that release will support CUDAToolkit.

**Comment by ngimel** [comment]
@malfet, @seemethere can we switch to CMake 3.17 given that?

**Comment by malfet** [comment]
We could start by vendoring CUDAToolkit.cmake in cmake/ folder. Also, to solve the problem we first need to update the detection in dependent libraries, don't we?

**Comment by peterbell10** [comment]
>We could start by vendoring `CUDAToolkit.cmake` in `cmake/` folder. 

Yes, that was my first though. The main problems with vendoring are:
- LibTorch users also need to have `FindCUDAToolkit.cmake` for `find_package(Torch)` to work. 
- dependencies need to independently find the same cuda library, without depending on the code vendored by PyTorch.

These are solvable problems, so if CMake 3.17 is a no-go then I can work with it. However, I'd rather not waste time if we could just upgrade CMake instead.

> to solve the problem we first need to update the detection in dependent libraries, don't we?

I'm looking into this.  `kineto` and `tensorpipe` are both under the PyTorch github org so I'm thinking we just change them to match PyTorch. `gloo` is under `facebookincubator`  so might be more of an issue, but I assume we will have some wiggle room there as well since it's still related.

Best case scenario, we can just move them over to modern CMake as well. Worst case, we might need to add a mode flag that changes how they search for CUDA libraries to match PyTorch. Either way, I expect it will be simpler if I can assume a newer cmake.

**Comment by janbernloehr** [comment]
Any update on this one? Did some local testing and it would also greatly simplify the use of compiler caches.

**Comment by peterbell10** [comment]
I'm working on vendoring/backporting `FindCUDATookit.cmake` in #82695. Unfortunately it seems some of the less common targets aren't well tested on windows, so it's not completely straight-forward.

**Comment by jakirkham** [comment]
Given some time has passed, am curious whether upgrading (without vendoring) is more approachable

**Comment by peterbell10** [comment]
This was pretty close to done. We already use a vendored version of `FindCUDAToolkit.cmake` but the `tensorpipe` submodule still uses `find_package(CUDA)`. I sent a PR to fix `tensorpipe` but it never got any response https://github.com/pytorch/tensorpipe/pull/454

**Comment by jakirkham** [comment]
Thanks Peter! 🙏

Interesting, it looks like tensorpipe is archived. Does PyTorch still use it? If so, is there a better source to modify at stage?

Also noticed these bits. Should these be removed as well?

https://github.com/pytorch/pytorch/blob/2ab8b344332c5a49913dc204ac3bb8248090359c/cmake/public/cuda.cmake#L28-L29

https://github.com/pytorch/pytorch/blob/2ab8b344332c5a49913dc204ac3bb8248090359c/cmake/Modules_CUDA_fix/upstream/FindCUDA.cmake#L2

**Comment by peterbell10** [comment]
That's pretty weird since we build with `tensorpipe` by default and the docs say it's the default rpc backend:
https://pytorch.org/docs/stable/rpc.html#tensorpipe-backend

@malfet @albanD @lw any idea why this would be archived if it's still actively used?

> Also noticed these bits. Should these be removed as well?

Once the dependencies are updated they can be removed. Until then we need these to make sure the two ways of finding CUDA libraries are actually pointing to the same libs.

**Comment by lw** [comment]
As far as I am aware the RPC module didn't get much traction thus, transitively, TensorPipe is also quite unused. This means that it didn't manage to "graduate" to become maintained by PyTorch and got abandoned. If it needs to be revived in order to unblock this change then unfortunately we'll first have to figure out its support model...

**Comment by peterbell10** [comment]
To me it sounds like it should be deprecated then, but archiving should really wait until it's removed form the main project. Otherwise we're going to be held back by its build system.

**Comment by jakirkham** [comment]
Continuing to use an archived project as a dependency sounds challenging for more than one reason. For example what if a CVE is found in that deprecated dependency?

**Comment by h-vetinari** [comment]
This is still being used (at least) in Caffe2
https://github.com/pytorch/pytorch/blob/bb4964013f860595b6e90cc964cc400d76990ed4/cmake/public/cuda.cmake#L29
which is included through torch, and ends up causing failures at CMake configure time

**Comment by albanD** [comment]
Ho we did move 3.18 so this should be feasible indeed!

**Comment by h-vetinari** [comment]
I opened a PR with the patch that we ended up using to fix this (out of necessity...) in conda-forge: https://github.com/pytorch/pytorch/pull/147300

It's a long way from being mergeable, but perhaps it helps as a first step.

**Comment by albanD** [comment]
Thanks for sending a PR to get this rolling!
Beyond the submodule upgrades that I guess are by mistake, the PR sounds reasonable. 

We're not doing this kind of deep cmake changes very often on our end, what do you think is the right process to make sure we don't break anything downstream when making such changes?

**Comment by h-vetinari** [comment]
@albanD, did you mean to comment on #147300? I just noticed the spurious submodule changes, will fix them now.

**Comment by albanD** [comment]
Ho yeah, I think I started writing this on the PR but then moved here since it felt more relevant haha

**Comment by Skylion007** [comment]
@cyyever This might be good to close out while we are cleanup the CMake
